### ISSUE TITLE
Support Brotli decompression

### ISSUE THREAD

**Issue by forresthopkinsa** [request/request#2866]
### Summary
This library doesn't seem to support a `br` content type

### Expected Behavior
When receiving a Brotli-compressed response, the library should be able to decompress it exactly the same way as a `gzip` response

### Current Behavior
It doesn't make any attempt to decode it.

### Context
This is causing problems in higher level apps, i.e. Postman: postmanlabs/postman-app-support#4091

**Comment by flyweight** [comment]
https://github.com/nodejs/node/pull/24938

**Comment by ZJONSSON** [comment]
Would be great to wait for the native brotli to be released.  The userland [brotly](https://www.npmjs.com/package/brotli) takes up 1.5 meg in node modules.

**Comment by flyweight** [comment]
For those of you that are waiting on this, here is a temporary workaround (with fallback to gzip where brotli isn't available from the server):

## Required Modules

https://www.npmjs.com/package/iltorb
https://www.npmjs.com/package/patch-package

## Diff for patch-package

```diff
$ cat patches/request+2.87.0.patch 
patch-package
--- a/node_modules/request/request.js
+++ b/node_modules/request/request.js
@@ -379,8 +379,12 @@ Request.prototype.init = function (options) {
     )
   }
 
+  if (typeof self.brotliDecompress !== 'function') {
+    delete self.brotliDecompress
+  }
+
   if (self.gzip && !self.hasHeader('accept-encoding')) {
-    self.setHeader('accept-encoding', 'gzip, deflate')
+    self.setHeader('accept-encoding', `${self.brotliDecompress ? 'br, ' : ''}gzip, deflate`)
   }
 
   if (self.uri.auth && !self.hasHeader('authorization')) {
@@ -1027,7 +1031,10 @@ Request.prototype.onRequestResponse = function (response) {
         finishFlush: zlib.Z_SYNC_FLUSH
       }
 
-      if (contentEncoding === 'gzip') {
+      if (contentEncoding === 'br') {
+        responseContent = self.brotliDecompress()
+        response.pipe(responseContent)
+      } else if (contentEncoding === 'gzip') {
         responseContent = zlib.createGunzip(zlibOptions)
         response.pipe(responseContent)
       } else if (contentEncoding === 'deflate') {
```

## Example Usage

```javascript
const { decompressStream } = require('iltorb')
const request = require('request')

request({
  uri: 'https://www.google.com',
  gzip: true,
  brotliDecompress: decompressStream,
  headers: {
    'user-agent': 'curl/7.47.0'
  }
})
```

**Comment by AugustGarcia** [comment]
Another workaround:
```
import brotli

# Make some request here

if (resp.headers["Content-Encoding"] == "br"):
    resp.decompressed_content = brotli.decompress(resp.content)
else:
    resp.decompressed_content = resp.content
```

**Comment by karlhorky** [comment]
Maybe @mikeal's comment here in https://github.com/request/request/issues/2907#issuecomment-533747342 also applies?

> request is in maintenance mode and we’re not accepting any new features. for what it’s worth, I did implement this in bent.

**Comment by forresthopkinsa** [comment]
@flyweight With brotli support long-since-landed in Node, how is this issue looking?

**Comment by cloudily** [comment]
For those with node versions that support `createBrotliDecompress`:

### patch-package
```diff
$ cat patches/request+2.87.0.patch 
patch-package
--- a/node_modules/request/request.js
+++ b/node_modules/request/request.js
@@ -380,7 +380,7 @@ Request.prototype.init = function (options) {
   }
 
   if (self.gzip && !self.hasHeader('accept-encoding')) {
-    self.setHeader('accept-encoding', 'gzip, deflate')
+    self.setHeader('accept-encoding', 'br, gzip, deflate')
   }
 
   if (self.uri.auth && !self.hasHeader('authorization')) {
@@ -1027,7 +1027,10 @@ Request.prototype.onRequestResponse = function (response) {
         finishFlush: zlib.Z_SYNC_FLUSH
       }
 
-      if (contentEncoding === 'gzip') {
+      if (contentEncoding === 'br') {
+        responseContent = zlib.createBrotliDecompress()
+        response.pipe(responseContent)
+      } else if (contentEncoding === 'gzip') {
         responseContent = zlib.createGunzip(zlibOptions)
         response.pipe(responseContent)
       } else if (contentEncoding === 'deflate') {
```

### Example

```javascript
const request = require('request')

request({
  uri: 'https://www.google.com',
  gzip: true,
  headers: {
    'user-agent': 'curl/7.47.0'
  }
})
```

**Comment by yamin8000** [comment]
I think this issue should be fixed and reopened, I'm using postman 8.7.0, when response is compressed in brotli, data is loaded correctly but size of data is equal to uncompressed data even though `Accept-Encoding` is set to `gzip, deflate, br` by default and in response headers `Content-Encoding` is equal to `br`. I tried same request using browser and in browser it's compressed and has lower size/content-length.
